Enhancing the scale-up detection conditions.

ishan16696 opened the pull request:
**What this PR does / why we need it**:
This PR improves the scale-up detection conditions to avoid potential failure while migrating single node etcd to multi-node etcd cluster.

**Which issue(s) this PR fixes**:
Fixes #606 

**Special notes for your reviewer**:


**Release note**:
```improvement operator
Enhances the scale-up detection conditions to avoid potential failure while from migrating single node etcd to multi-node etcd cluster.
```

seshachalam-yv commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -564,3 +566,12 @@ func GetPrevScheduledSnapTime(nextSnapSchedule time.Time, timeWindow float64) ti
 	)
 }
+
+// GetBackoff returns the backoff with Factor=2
+func GetBackoff(retryPeriod time.Duration, steps int) wait.Backoff {
+	backoff := retry.DefaultBackoff
+	backoff.Duration = retryPeriod
+	backoff.Factor = 2
+	backoff.Steps = steps
+	return backoff
+}
```
```suggestion
// GetBackoff returns an exponential backoff with a factor of 2. The retry period and maximum number of steps for the backoff
// are specified by the `retryPeriod` and `steps` arguments, respectively
func GetBackoff(retryPeriod time.Duration, steps int) wait.Backoff {
	return wait.Backoff{
		Steps:    steps,
		Duration: retryPeriod,
		Factor:   2,
		Jitter:   retry.DefaultBackoff.Jitter,
	}
}
```

seshachalam-yv commented on pkg/member/member_control.go:
```diff
@@ -299,3 +303,30 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+		return true, nil
+	}
+
+	isEtcdMemberPresent, err := m.IsMemberInCluster(ctx)
+	if err != nil {
+		return false, err
+	}
+	return !isEtcdMemberPresent, nil
```
```suggestion
	isEtcdMemberPresent, err := m.IsMemberInCluster(ctx)
	if err != nil || isEtcdMemberPresent {
		return false, err
	}
	return true, nil
```

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -299,3 +303,30 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Fatalf("failed to create clientset, %v", err)
```
yes, it is

unmarshall replied:
I agree with Sesha, that we should not cause an exit from this method. This method should return an error. The caller of this method can decide to terminate instead. It also makes writing tests for this easier.

unmarshall commented on pkg/member/member_control.go:
```diff
@@ -299,3 +303,30 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
 }
+
+// IsScaleup determines whether a etcd cluster is getting scale-up or not.
+func IsScaleup(ctx context.Context, m Control) (bool, error) {
```
Can we do the following:

1. Rename this method to `clusterScaled`
2. Make it a method in `Control` interface instead of passing an instance of `Control` to it.
3. This method does 2 completely different things which do not belong in a single method:
    - It checks if the cluster is scaled from nonHA to HA by calling `miscellaneous.GetInitialClusterStateIfScaleup`
    - It check if the current member is already in cluster.

ishan16696 replied:
> This method does 2 completely different things which do not belong in a single method:
It checks if the cluster is scaled from nonHA to HA by calling miscellaneous.GetInitialClusterStateIfScaleup
It check if the current member is already in cluster.

What makes you think that ?
IMO it `IsScaleup` does only 1 thing which it checks cluster is getting scaled-up or not using these 2 function calls:
`miscellaneous.GetInitialClusterStateIfScaleup` and `IsMemberInCluster`.

If `miscellaneous.GetInitialClusterStateIfScaleup`  says I have the annotation in sts --> surely it is scaleup case return true no need to call `IsMemberInCluster`

If `miscellaneous.GetInitialClusterStateIfScaleup` says I don't have annotation in sts  --> we still not sure about it is not being scaleup case due to bug in adding annotation to sts: https://github.com/gardener/etcd-druid/issues/557

ishan16696 replied:
>  Rename this method to clusterScaled

what about `IsClusterScaledup` ?

unmarshall replied:
>what about IsClusterScaledup ?

Maybe we can use `IsClusterScaledUp`

unmarshall commented on pkg/member/member_control.go:
```diff
@@ -299,3 +303,30 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+	logger := logrus.New().WithField("actor", "scale-up")
+	// fetch pod name from env
+	podNS := os.Getenv("POD_NAMESPACE")
+	podName := os.Getenv("POD_NAME")
```
instead use `memberControl.podName`.

unmarshall commented on pkg/member/member_control.go:
```diff
@@ -299,3 +303,30 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+func IsScaleup(ctx context.Context, m Control) (bool, error) {
+	logger := logrus.New().WithField("actor", "scale-up")
+	// fetch pod name from env
+	podNS := os.Getenv("POD_NAMESPACE")
```
make `podNamespace` as an additional field in `memberControl` struct and access that.

unmarshall commented on pkg/initializer/initializer.go:
```diff
@@ -54,11 +54,9 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
-			}, func() error {
+		isScaleup, err := member.IsScaleup(ctx, m)
+		if isScaleup && err == nil {
+			retry.OnError(retry.DefaultBackoff, errors.AnyError, func() error {
```
if one is sure that a cluster has been scaled up from nonHA to HA then there can be 2 approaches:
1. Infinitely attempt to add a member as learner till it succeeds
2. Try for fixed number of times and then exit causing a restart of backup-restore.
One should not proceed ahead without adding this member as a learner.

unmarshall commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -564,3 +566,12 @@ func GetPrevScheduledSnapTime(nextSnapSchedule time.Time, timeWindow float64) ti
 }
+
+// GetBackoff returns the backoff with Factor=2
+func GetBackoff(retryPeriod time.Duration, steps int) wait.Backoff {
```
```suggestion
func CreateBackoff(retryPeriod time.Duration, steps int) wait.Backoff {
```

unmarshall commented on pkg/member/member_control.go:
```diff
@@ -38,6 +42,9 @@ type Control interface {
 	AddMemberAsLearner(context.Context) error
 
+	// IsClusterScaledup determines whether a etcd cluster is getting scale-up or not.
+	IsClusterScaledup(context.Context, client.Client) (bool, error)
```
Scaledup are 2 words and therefore should be camel-cased. So the name should be `IsClusterScaledUp`

ishan16696 commented:
I have addressed reviews comments @unmarshall @seshachalam-yv . PTAL
I haven’t added unit tests. I will add, currently I'm running some tests on this PR.

unmarshall commented on pkg/initializer/initializer.go:
```diff
@@ -45,23 +45,29 @@ import (
-				return err != nil
-			}, func() error {
+		isScaleup, err := m.IsClusterScaledup(ctx, clientSet)
+		if isScaleup && err == nil {
```
if a call to `IsClusterScaledup` returns an error then you currently assume that it is not a scaleup case and completely skip `AddMemberAsLeaner`.

unmarshall replied:
>I found that if IsClusterScaledup returns an error we have to assume that it is not a scaleup case else bootstrap case 0->3 won't be possible.

There is always a possibility that an error is returned even in the case of a scale-up when calling `IsMemberInCluster` - therefore assuming that if there is an error therefore it is not a scale up case but a bootstrap (0-3) case is IMHO incorrect.

aaronfern replied:
Its very difficult to distinguish between the errors here
For instance, `IsMemberInCluster` internally makes a call to `ProbeEtcd`. This could fail because we're in the bootstrapping case and etcds are not running, or due to some network issue/other issue. It's very difficult to deterministically differentiate between these two right now

aaronfern commented on pkg/member/member_control.go:
```diff
@@ -299,3 +314,24 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+
+	state, err := miscellaneous.GetInitialClusterStateIfScaleup(ctx, m.logger, clientSet, m.podName, m.podNamespace)
+	if err != nil {
+		m.logger.Errorf("unable to check scale-up case :%v", err)
```
This log is incorrect here, as we can print this message and then return `true` further down
Maybe we can update this message or remove this log altogether

aaronfern commented on pkg/member/member_control.go:
```diff
@@ -18,12 +18,16 @@ import (
 	"go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
 	"go.etcd.io/etcd/etcdserver/etcdserverpb"
 	"k8s.io/client-go/util/retry"
+	//"sigs.k8s.io/controller-runtime/pkg/client"
```
Remove this commented import

aaronfern commented on pkg/member/member_control.go:
```diff
@@ -8,7 +8,7 @@ import (
 
 	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
-	"github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
+	etcdClient "github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
```
```suggestion
	etcdclient "github.com/gardener/etcd-backup-restore/pkg/etcdutil/client"
```

ishan16696 replied:
why not follow camel-case ?

unmarshall replied:
`etcdClient` is just fine

seshachalam-yv commented on pkg/member/member_control.go:
```diff
@@ -299,3 +314,24 @@ func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+
+	isEtcdMemberPresent, err := m.IsMemberInCluster(ctx)
+	if err != nil || isEtcdMemberPresent {
+		return false, err
```
We are returning an error, but it is not being handled [here](https://github.com/gardener/etcd-backup-restore/pull/608/files#diff-dc0aa715ea97877db344484988dcdffda1fc0eaeda89257be6c70e9d724721d7R58).
[`IsMemberInCluster`](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/member/member_control.go#L125-L172) is returning multiple errors. 
To ensure proper error handling, we should either handle the error or log it with a reason for why we are ignoring it.

ishan16696 replied:
I have log the error. PTAL

seshachalam-yv commented on pkg/initializer/initializer.go:
```diff
@@ -45,24 +45,29 @@ import (
-			retry.OnError(retry.DefaultBackoff, func(err error) bool {
-				return err != nil
-			}, func() error {
+		isScaleup, err := m.IsClusterScaledUp(ctx, clientSet)
```
```go
		isScaleUp, err := m.IsClusterScaledUp(ctx, clientSet)
		if err != nil {
			logger.Warnf("Failed to verify cluster scale-up: %v. Proceeding with bootstrap operation.", err)
		}
		if isScaleup && err == nil {
			err = retry.OnError(retry.DefaultBackoff, errors.AnyError, func() error {
				return m.AddMemberAsLearner(ctx)
			})
			if err != nil {
				logger.Errorf("adding learner member failed: %v", err)
				return err
			}
			// Non-voting member (learner) has been successfully added, so no further restoration or validation is needed.
			return nil
		}
```
@ishan16696 
**Nit:** please check error first

seshachalam-yv replied:
WDYT `logger.Warnf("Failed to verify cluster scale-up: %v. Proceeding with bootstrap operation.", err)`

ishan16696 replied:
I will address it as a separate PR as I also need to add need unit test. I hope it make sense.

seshachalam-yv reviewed (commented):
Apart from the NIT item, LGTM.
/lgtm
